x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/chunked/ChunkedDataExtractor.java

```diff
@@ -180,7 +180,7 @@ private Result getNextStream() throws IOException {
     private void advanceTime() {
         currentStart = currentEnd;
         currentEnd = Math.min(currentStart + chunkSpan, context.end);
-        currentExtractor = dataExtractorFactory.newExtractor(currentStart, currentEnd);
+        currentExtractor = dataExtractorFactory.newExtractor(currentStart, currentEnd, context.query);
```

Comment:
I think you've found the line with the bug - well done.

But I think the fix is going to lead to the query getting bigger and bigger with each chunk.

For each new chunk the query from the datafeed config will get added in again. Functionally it won't matter (because `x > 7 && !cold_or_frozen` gives the same answer as `x > 7 && x > 7 && !cold_or_frozen`). But by the 10th chunk we'll be up to `x > 7 && x > 7 && x > 7 && x > 7 && x > 7 && x > 7 && x > 7 && x > 7 && x > 7 && x > 7 && !cold_or_frozen` and that will cause complaints if people see that query in logs or task lists.

So we probably need a new overload of `newExtractor` where the query supplied is the full query rather than just an extra bit.

Reply:
I don't think that's the case: if I understand correctly every time a completely new data extractor is created from `currentStart`, `currentEnd` and `context.query`, and these inputs are left untouched.

Reply:
You're right, it doesn't _recursively_ append the same query over and over again. The outer chunked data extractor's context doesn't change.

However, it _does_ append the datafeed query twice, for example:

```
[2023-11-23T13:55:54,253][INFO ][o.e.x.m.d.e.s.ScrollDataExtractor] [runTask-0] About to run search {"size":1000,"query":{"bool":{"filter":[{"bool":{"filter":[{"bool":{"filter":[{"term":{"Carrier":{"value":"Kibana Airlines"}}}],"boost":1.0}},{"bool":{"filter":[{"bool":{"filter":[{"term":{"Carrier":{"value":"Kibana Airlines"}}}],"boost":1.0}},{"bool":{"must_not":[{"terms":{"_tier":["data_frozen","data_cold"],"boost":1.0}}],"boost":1.0}}],"boost":1.0}}],"boost":1.0}},{"range":{"timestamp":{"gte":1699833600000,"lt":1711104839515,"format":"epoch_millis","boost":1.0}}}],"boost":1.0}},"_source":false,"stored_fields":"_none_","docvalue_fields":[{"field":"timestamp","format":"epoch_millis"}],"sort":[{"timestamp":{"order":"asc"}}]}
```

In that the datafeed config query is `{"bool":{"filter":[{"term":{"Carrier":{"value":"Kibana Airlines"}}}]}}`. It could be something much bigger, that would be very conspicuous if duplicated.

I got that by adding this line as the very first line of `ScrollDataExtractor.executeSearchRequest`:

```
    logger.info("About to run search " + searchRequestBuilder);
```

I was using the Kibana sample flight data that you can get by clicking "Try sample data" on the Kibana home page.

So I still think it's worth having a new method that the `ChunkedDataExtractor` can call on the `DataExtractorFactory` it wraps to tell it that it's providing the full query and not one that needs to have the query from the datafeed config added into it.

Reply:
The search query for the preview endpoint now looks like:
```
{
  "size": 1000,
  "query": {
    "bool": {
      "filter": [
        {
          "bool": {
            "filter": [
              {
                "bool": {
                  "must_not": [
                    {
                      "term": {
                        "bla": {
                          "value": "bla"
                        }
                      }
                    }
                  ],
                  "boost": 1.0
                }
              },
              {
                "bool": {
                  "must_not": [
                    {
                      "terms": {
                        "_tier": [
                          "data_frozen",
                          "data_cold"
                        ],
                        "boost": 1.0
                      }
                    }
                  ],
                  "boost": 1.0
                }
              }
            ],
            "boost": 1.0
          }
        },
        {
          "range": {
            "@timestamp": {
              "gte": 1700611260000,
              "lt": 9223372036854775807,
              "format": "epoch_millis",
              "boost": 1.0
            }
          }
        }
      ],
      "boost": 1.0
    }
  },
  "_source": false,
  "stored_fields": "_none_",
  "docvalue_fields": [
    {
      "field": "@timestamp",
      "format": "epoch_millis"
    },
    {
      "field": "id"
    }
  ],
  "sort": [
    {
      "@timestamp": {
        "order": "asc"
      }
    }
  ]
}
```